File: ai-ml/computer-vision.js
/**
 * Computer Vision - JavaScript
 * 
 * Processing and understanding images:
 * - Image filtering
 * - Edge detection (simplified)
 * - Object detection (conceptual)
 */

// Image representation (simplified - 2D array of pixels)
class Image {
    constructor(width, height, data = null) {
        this.width = width;
        this.height = height;
        this.data = data || Array(height).fill(0).map(() => Array(width).fill(0));
    }
    
    getPixel(x, y) {
        if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
            return 0;
        }
        return this.data[y][x];
    }
    
    setPixel(x, y, value) {
        if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
            this.data[y][x] = Math.max(0, Math.min(255, value));
        }
    }
}

// Image Filtering
class ImageFilter {
    // Gaussian blur (simplified)
    static gaussianBlur(image, kernelSize = 3) {
        const kernel = this.createGaussianKernel(kernelSize);
        const result = new Image(image.width, image.height);
        
        const offset = Math.floor(kernelSize / 2);
        for (let y = 0; y < image.height; y++) {
            for (let x = 0; x < image.width; x++) {
                let sum = 0;
                for (let ky = 0; ky < kernelSize; ky++) {
                    for (let kx = 0; kx < kernelSize; kx++) {
                        const px = x + kx - offset;
                        const py = y + ky - offset;
                        sum += image.getPixel(px, py) * kernel[ky][kx];
                    }
                }
                result.setPixel(x, y, sum);
            }
        }
        return result;
    }
    
    static createGaussianKernel(size) {
        const kernel = Array(size).fill(0).map(() => Array(size).fill(0));
        const sigma = size / 3;
        const center = Math.floor(size / 2);
        let sum = 0;
        
        for (let y = 0; y < size; y++) {
            for (let x = 0; x < size; x++) {
                const dx = x - center;
                const dy = y - center;
                const value = Math.exp(-(dx * dx + dy * dy) / (2 * sigma * sigma));
                kernel[y][x] = value;
                sum += value;
            }
        }
        
        // Normalize
        for (let y = 0; y < size; y++) {
            for (let x = 0; x < size; x++) {
                kernel[y][x] /= sum;
            }
        }
        
        return kernel;
    }
    
    // Edge detection (Sobel operator - simplified)
    static sobelEdgeDetection(image) {
        const sobelX = [[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]];
        const sobelY = [[-1, -2, -1], [0, 0, 0], [1, 2, 1]];
        const result = new Image(image.width, image.height);
        
        for (let y = 1; y < image.height - 1; y++) {
            for (let x = 1; x < image.width - 1; x++) {
                let gx = 0, gy = 0;
                
                for (let ky = -1; ky <= 1; ky++) {
                    for (let kx = -1; kx <= 1; kx++) {
                        const pixel = image.getPixel(x + kx, y + ky);
                        gx += pixel * sobelX[ky + 1][kx + 1];
                        gy += pixel * sobelY[ky + 1][kx + 1];
                    }
                }
                
                const magnitude = Math.sqrt(gx * gx + gy * gy);
                result.setPixel(x, y, magnitude);
            }
        }
        
        return result;
    }
}

// Object Detection (Conceptual - simplified)
class ObjectDetector {
    constructor() {
        this.templates = {}; // Stored templates/features
    }
    
    // Template matching (simplified)
    detectObjects(image, template) {
        const matches = [];
        const threshold = 0.8;
        
        for (let y = 0; y <= image.height - template.height; y++) {
            for (let x = 0; x <= image.width - template.width; x++) {
                const similarity = this.calculateSimilarity(image, template, x, y);
                if (similarity > threshold) {
                    matches.push({ x, y, similarity });
                }
            }
        }
        
        return matches;
    }
    
    calculateSimilarity(image, template, startX, startY) {
        let sum = 0;
        let templateSum = 0;
        let imageSum = 0;
        
        for (let y = 0; y < template.height; y++) {
            for (let x = 0; x < template.width; x++) {
                const imgPixel = image.getPixel(startX + x, startY + y);
                const templatePixel = template.getPixel(x, y);
                sum += imgPixel * templatePixel;
                templateSum += templatePixel * templatePixel;
                imageSum += imgPixel * imgPixel;
            }
        }
        
        const denominator = Math.sqrt(templateSum * imageSum);
        return denominator > 0 ? sum / denominator : 0;
    }
}

// Example usage
console.log('=== Computer Vision ===\n');

// Create a simple test image (gradient)
console.log('Image Processing:');
const image = new Image(10, 10);
for (let y = 0; y < 10; y++) {
    for (let x = 0; x < 10; x++) {
        image.setPixel(x, y, (x + y) * 10);
    }
}

console.log('Original image (sample):');
for (let y = 0; y < 3; y++) {
    const row = [];
    for (let x = 0; x < 3; x++) {
        row.push(image.getPixel(x, y));
    }
    console.log(row.join(' '));
}

// Apply Gaussian blur
const blurred = ImageFilter.gaussianBlur(image, 3);
console.log('\nBlurred image (sample):');
for (let y = 0; y < 3; y++) {
    const row = [];
    for (let x = 0; x < 3; x++) {
        row.push(Math.round(blurred.getPixel(x, y)));
    }
    console.log(row.join(' '));
}

// Edge detection
const edges = ImageFilter.sobelEdgeDetection(image);
console.log('\nEdge detection (sample):');
for (let y = 1; y < 4; y++) {
    const row = [];
    for (let x = 1; x < 4; x++) {
        row.push(Math.round(edges.getPixel(x, y)));
    }
    console.log(row.join(' '));
}

module.exports = { Image, ImageFilter, ObjectDetector };

File: ai-ml/computer-vision.py
"""
Computer Vision - Python

Processing and understanding images:
- Image filtering
- Edge detection (simplified)
- Object detection (conceptual)
"""

import math
from typing import List, Tuple, Dict
import numpy as np


# Image representation (simplified - 2D array of pixels)
class Image:
    def __init__(self, width: int, height: int, data: List[List[int]] = None):
        self.width = width
        self.height = height
        self.data = data if data else [[0] * width for _ in range(height)]
    
    def get_pixel(self, x: int, y: int) -> int:
        if x < 0 or x >= self.width or y < 0 or y >= self.height:
            return 0
        return self.data[y][x]
    
    def set_pixel(self, x: int, y: int, value: int):
        if 0 <= x < self.width and 0 <= y < self.height:
            self.data[y][x] = max(0, min(255, int(value)))


# Image Filtering
class ImageFilter:
    @staticmethod
    def gaussian_blur(image: Image, kernel_size: int = 3) -> Image:
        kernel = ImageFilter.create_gaussian_kernel(kernel_size)
        result = Image(image.width, image.height)
        
        offset = kernel_size // 2
        for y in range(image.height):
            for x in range(image.width):
                total = 0
                for ky in range(kernel_size):
                    for kx in range(kernel_size):
                        px = x + kx - offset
                        py = y + ky - offset
                        total += image.get_pixel(px, py) * kernel[ky][kx]
                result.set_pixel(x, y, total)
        return result
    
    @staticmethod
    def create_gaussian_kernel(size: int) -> List[List[float]]:
        kernel = [[0] * size for _ in range(size)]
        sigma = size / 3
        center = size // 2
        total = 0
        
        for y in range(size):
            for x in range(size):
                dx = x - center
                dy = y - center
                value = math.exp(-(dx * dx + dy * dy) / (2 * sigma * sigma))
                kernel[y][x] = value
                total += value
        
        # Normalize
        for y in range(size):
            for x in range(size):
                kernel[y][x] /= total
        
        return kernel
    
    @staticmethod
    def sobel_edge_detection(image: Image) -> Image:
        sobel_x = [[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]]
        sobel_y = [[-1, -2, -1], [0, 0, 0], [1, 2, 1]]
        result = Image(image.width, image.height)
        
        for y in range(1, image.height - 1):
            for x in range(1, image.width - 1):
                gx = gy = 0
                
                for ky in range(-1, 2):
                    for kx in range(-1, 2):
                        pixel = image.get_pixel(x + kx, y + ky)
                        gx += pixel * sobel_x[ky + 1][kx + 1]
                        gy += pixel * sobel_y[ky + 1][kx + 1]
                
                magnitude = math.sqrt(gx * gx + gy * gy)
                result.set_pixel(x, y, magnitude)
        
        return result


# Object Detection (Conceptual - simplified)
class ObjectDetector:
    def __init__(self):
        self.templates: Dict = {}
    
    def detect_objects(self, image: Image, template: Image) -> List[Dict]:
        matches = []
        threshold = 0.8
        
        for y in range(image.height - template.height + 1):
            for x in range(image.width - template.width + 1):
                similarity = self.calculate_similarity(image, template, x, y)
                if similarity > threshold:
                    matches.append({'x': x, 'y': y, 'similarity': similarity})
        
        return matches
    
    def calculate_similarity(self, image: Image, template: Image, start_x: int, start_y: int) -> float:
        total = 0
        template_sum = 0
        image_sum = 0
        
        for y in range(template.height):
            for x in range(template.width):
                img_pixel = image.get_pixel(start_x + x, start_y + y)
                template_pixel = template.get_pixel(x, y)
                total += img_pixel * template_pixel
                template_sum += template_pixel * template_pixel
                image_sum += img_pixel * img_pixel
        
        denominator = math.sqrt(template_sum * image_sum)
        return total / denominator if denominator > 0 else 0


# Example usage
if __name__ == "__main__":
    print("=== Computer Vision ===\n")
    
    # Create a simple test image (gradient)
    print("Image Processing:")
    image = Image(10, 10)
    for y in range(10):
        for x in range(10):
            image.set_pixel(x, y, (x + y) * 10)
    
    print("Original image (sample):")
    for y in range(3):
        row = [str(image.get_pixel(x, y)) for x in range(3)]
        print(" ".join(row))
    
    # Apply Gaussian blur
    blurred = ImageFilter.gaussian_blur(image, 3)
    print("\nBlurred image (sample):")
    for y in range(3):
        row = [str(int(blurred.get_pixel(x, y))) for x in range(3)]
        print(" ".join(row))
    
    # Edge detection
    edges = ImageFilter.sobel_edge_detection(image)
    print("\nEdge detection (sample):")
    for y in range(1, 4):
        row = [str(int(edges.get_pixel(x, y))) for x in range(1, 4)]
        print(" ".join(row))

